src/main/java/com/example/onlinebookshop/services/impl/ShoppingCartServiceImpl.java

```diff
@@ -0,0 +1,99 @@
+    }
+
+    @Override
+    public ShoppingCartDto getShoppingCartByUserEmail(Long userId) {
+        return shoppingCartMapper.toShoppingCartDto(shoppingCartRepository.findByUserId(userId));
+    }
+
+    @Override
+    public ShoppingCartDto addBookToShoppingCart(Long userId,
+                                                 AddCartItemDto addCartItemDto) {
+        Book book = bookRepository.findById(addCartItemDto.bookId()).orElseThrow(() ->
+                new EntityNotFoundException("Book with id " + addCartItemDto.bookId()
+                        + " not found"));
+        ShoppingCart shoppingCart = shoppingCartRepository.findByUserId(userId);
+        shoppingCart.getCartItems().stream()
+                .filter(item -> item.getBook().getId().equals(addCartItemDto.bookId()))
+                .findFirst()
+                .ifPresentOrElse(item -> {
+                    if (item.getQuantity() + addCartItemDto.quantity() <= 100) {
+                        item.setQuantity(item.getQuantity() + addCartItemDto.quantity());
+                    } else {
+                        throw new TooManyObjectsException(QUANTITY_DESCRIPTION);
+                    }
+                },() -> addCartItemToCart(addCartItemDto, book, shoppingCart));
+        shoppingCartRepository.save(shoppingCart);
+        return shoppingCartMapper.toShoppingCartDto(shoppingCart);
+    }
+
+    @Override
+    public ShoppingCartDto updateBookQuantity(Long userId, Long cartItemId,
+                                              UpdateItemQuantityDto itemQuantityDto) {
+        ShoppingCart shoppingCart = shoppingCartRepository.findByUserId(userId);
+        CartItem cartItem = cartItemRepository.findByIdAndShoppingCartId(
+                        cartItemId, shoppingCart.getId())
+                .map(item -> {
+                    item.setQuantity(itemQuantityDto.quantity());
+                    return item;
+                }).orElseThrow(() -> new EntityNotFoundException(
+                        String.format("No cart item with id: %d for user: %d", cartItemId, userId)
+                ));
+        cartItemRepository.save(cartItem);
+        return shoppingCartMapper.toShoppingCartDto(shoppingCart);
+    }
+
+    @Override
+    public void deleteBookFromShoppingCart(Long userId, Long cartItemId) {
+        ShoppingCart shoppingCart = shoppingCartRepository.findByUserId(userId);
+        CartItem cartItem = cartItemRepository.findByIdAndShoppingCartId(
+                cartItemId, shoppingCart.getId()).orElseThrow(() ->
+                new EntityNotFoundException(("Cart item with id " + cartItemId + " not found")));
+        shoppingCart.getCartItems().remove(cartItem);
+        cartItemRepository.delete(cartItem);
+        shoppingCartRepository.save(shoppingCart);
+    }
+
+    private void addCartItemToCart(AddCartItemDto addCartItemDto, Book book,
+                                   ShoppingCart shoppingCart) {
+        CartItem cartItem = new CartItem();
```

Comment:
Or you can use 
```suggestion
        CartItem cartItem = itemMapper.toCartItem(addCartItemDto);
        cartItem.setBook(book);
        shoppingCart.addItemToCart(cartItem);
```

Where in ShoppingCart entity 
```
public void addItemToCart(CartItem cartItem) {
        cartItem.setShoppingCart(this);
        cartItems.add(cartItem);
    }

```